components/layout_2020/display_list/stacking_context.rs

```diff
@@ -0,0 +1,369 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::display_list::DisplayListBuilder;
+use crate::fragments::{AnonymousFragment, BoxFragment, Fragment};
+use crate::geom::{PhysicalRect, ToWebRender};
+use gfx_traits::{combine_id_with_fragment_type, FragmentType};
+use std::cmp::Ordering;
+use std::mem;
+use style::computed_values::float::T as ComputedFloat;
+use style::computed_values::mix_blend_mode::T as ComputedMixBlendMode;
+use style::computed_values::overflow_x::T as ComputedOverflow;
+use style::computed_values::position::T as ComputedPosition;
+use style::computed_values::transform_style::T as ComputedTransformStyle;
+use style::values::computed::Length;
+use webrender_api::units::LayoutVector2D;
+use webrender_api::{ExternalScrollId, ScrollSensitivity, SpaceAndClipInfo, SpatialId};
+
+pub(crate) struct StackingContextFragment<'a> {
+    space_and_clip: SpaceAndClipInfo,
+    containing_block: PhysicalRect<Length>,
+    fragment: &'a Fragment,
+}
+
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub(crate) enum StackingContextType {
+    Real,
+    PseudoPositioned,
+    PseudoFloat,
+}
+
+pub(crate) struct StackingContext<'a> {
+    /// The type of this StackingContext. Used for collecting and sorting.
+    context_type: StackingContextType,
+
+    /// The `z-index` for this stacking context.
+    pub z_index: i32,
+
+    /// Fragments that make up the content of this stacking context.
+    fragments: Vec<StackingContextFragment<'a>>,
+
+    /// All non-float stacking context and pseudo stacking context children
+    /// of this stacking context.
+    stacking_contexts: Vec<StackingContext<'a>>,
+
+    /// All float pseudo stacking context children of this stacking context.
+    float_stacking_contexts: Vec<StackingContext<'a>>,
+}
+
+impl<'a> StackingContext<'a> {
+    pub(crate) fn new(context_type: StackingContextType, z_index: i32) -> Self {
+        Self {
+            context_type,
+            z_index,
+            fragments: vec![],
+            stacking_contexts: vec![],
+            float_stacking_contexts: vec![],
+        }
+    }
+
+    pub(crate) fn sort_stacking_contexts(&mut self) {
+        self.stacking_contexts.sort_by(|a, b| {
+            if a.z_index != 0 || b.z_index != 0 {
+                return a.z_index.cmp(&b.z_index);
+            }
+
+            match (a.context_type, b.context_type) {
```

Comment:
Instead of this enum being part of the sort, what do you think of having separate `Vec`s for each categories? (I’m not saying we should do that, I haven’t thought through the advantages or downsides of each approach.)

Reply:
So it's essentially a tradeoff between sorting one slightly larger Vec of stacking contexts versus sorting three or four smaller ones.  I'm not sure which is better, but I've tried to more-or-less copy the design of the old layout system which went through a process of iteration and optimization.